gateway/src/main/java/io/zeebe/gateway/impl/job/CreateWorkflowHandler.java

```diff
@@ -0,0 +1,95 @@
+import io.zeebe.gateway.cmd.BrokerErrorException;
+import io.zeebe.gateway.impl.broker.BrokerClient;
+import io.zeebe.gateway.impl.broker.BrokerResponseConsumer;
+import io.zeebe.gateway.impl.broker.RoundRobinDispatchStrategy;
+import io.zeebe.gateway.impl.broker.cluster.BrokerTopologyManager;
+import io.zeebe.gateway.impl.broker.request.BrokerCreateWorkflowInstanceRequest;
+import io.zeebe.protocol.impl.record.value.workflowinstance.WorkflowInstanceCreationRecord;
+import io.zeebe.protocol.record.ErrorCode;
+import io.zeebe.transport.impl.sender.NoRemoteAddressFoundException;
+import java.util.function.Consumer;
+
+public class CreateWorkflowHandler {
+
+  private final BrokerClient brokerClient;
+  private final RoundRobinDispatchStrategy roundRobinDispatchStrategy;
+
+  public CreateWorkflowHandler(
+      final BrokerClient brokerClient, final BrokerTopologyManager topologyManager) {
+    this.brokerClient = brokerClient;
+    roundRobinDispatchStrategy = new RoundRobinDispatchStrategy(topologyManager);
+  }
+
+  public void createWorkflow(
+      final int partitionsCount,
+      final BrokerCreateWorkflowInstanceRequest request,
+      final BrokerResponseConsumer<WorkflowInstanceCreationRecord> responseConsumer,
+      final Consumer<Throwable> throwableConsumer) {
+    createWorkflow(
+        request,
+        partitionIdIteratorForType(partitionsCount),
+        responseConsumer,
+        throwableConsumer,
+        null);
+  }
+
+  private void createWorkflow(
+      final BrokerCreateWorkflowInstanceRequest request,
+      final PartitionIdIterator partitionIdIterator,
+      final BrokerResponseConsumer<WorkflowInstanceCreationRecord> responseConsumer,
+      final Consumer<Throwable> throwableConsumer,
+      final Throwable lastError) {
+
+    if (partitionIdIterator.hasNext()) {
+      final int partitionId = partitionIdIterator.next();
+
+      // partitions to check
+      request.setPartitionId(partitionId);
+      brokerClient.sendRequest(
+          request,
+          responseConsumer,
+          error -> {
+            if (shouldRetry(error)) {
+              Loggers.GATEWAY_LOGGER.trace(
+                  "Failed to create workflow on partition {}",
+                  partitionIdIterator.getCurrentPartitionId(),
+                  error);
+              createWorkflow(
+                  request, partitionIdIterator, responseConsumer, throwableConsumer, error);
+            } else {
+              throwableConsumer.accept(error);
+            }
+          },
+          response -> false);
+    } else {
+      // no partition left to check
+      throwableConsumer.accept(lastError);
+    }
+  }
+
+  private boolean shouldRetry(final Throwable error) {
+    if (error instanceof NoRemoteAddressFoundException) {
+      return true;
+    } else if (error instanceof BrokerErrorException) {
+      return ((BrokerErrorException) error).getError().getCode()
+          == ErrorCode.PARTITION_LEADER_MISMATCH;
```

Comment:
Let's revisit the whitelist here for the 0.24 implementation, because there may be more than resource exhausted required (e.g. when upgrading maybe you want have different broker versions, so you might want to try a different partition on `INVALID_CLIENT_VERSION`?). When we write the issue for it, we should make sure not to forget about it.